src/gt4py/cartesian/gtc/passes/gtir_upcaster.py

```diff
@@ -52,22 +50,30 @@ def _numpy_ufunc_upcasting_rule(*dtypes, ufunc: np.ufunc):
     implementing C/C++ type promotion rules, restricted to supported types.
     See https://numpy.org/doc/stable/user/basics.ufuncs.html?highlight=index#type-casting-rules for details.
     """
+    matched: dict[int, list[common.DataType]] = {}
     for t in ufunc.types:
         inputs, output = t.split("->")
         assert len(inputs) == len(dtypes)
         if all(
-            arg_dtype <= typestr_to_data_type(np.dtype(cand_typestr).str)
+            arg_dtype <= common.typestr_to_data_type(np.dtype(cand_typestr).str)
             for arg_dtype, cand_typestr in zip(dtypes, inputs)
         ):
-            assert typestr_to_data_type(np.dtype(output).str) != DataType.INVALID
-            return [typestr_to_data_type(np.dtype(cand_typestr).str) for cand_typestr in inputs]
+            assert common.typestr_to_data_type(np.dtype(output).str) != common.DataType.INVALID
+            result = [
+                common.typestr_to_data_type(np.dtype(cand_typestr).str) for cand_typestr in inputs
+            ]
+            matched[sum(result)] = result
+    # take the one with the lowest precision, for binary functions with mixed parameter types this is a bit arbitrary,
+    # but it seems we don't have any of these functions anyway
+    return matched[min(matched.keys())]
```

Comment:
Maybe using `sorted()` would be simpler to understand here:
```suggestion
    return sorted(matched.items())[0]
```

Reply:
as discussed: leave it like it is